### PR TITLE
perf: replace zip_eq with zip_eq_fast/zip_eq_debug

### LINKED ISSUE #7779: replace `zip_eq` with optimizable version
As mentioned by @BugenZhao : https://github.com/risingwavelabs/risingwave/pull/7749#issuecomment-1422133394

And previously caught by @wangrunji0408 in https://github.com/risingwavelabs/risingwave/pull/6856.

We should probably replace `zip_eq` system wide with something else. This should be optimized away in release and bench builds.

### PR DESCRIPTION
**Pull request by xxchan**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?
close #7779

See the rustdoc in `src/common/src/util/iter_util.rs` for motivation.

## Checklist

- [x] I have written necessary rustdoc comments
- ~~[ ] I have added necessary unit tests and integration tests~~
- ~~[ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features).~~
- ~~[ ] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer the issue)~~
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

### CODE REVIEW

File: src/common/src/util/iter_util.rs [hunk 1]
```diff
@@ -0,0 +1,56 @@
+// Copyright 2023 RisingWave Labs
```
Comment (xxchan) [inline src/common/src/util/iter_util.rs:1]: here is the impl!

File: src/common/src/util/iter_util.rs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    B::IntoIter: ExactSizeIterator,
+{
+    /// A specialized version of `zip_eq` for [`ExactSizeIterator`].
+    ///
+    /// It's a separate trait because Rust doesn't support specialization yet.
+    /// See [tracking issue for specialization (RFC 1210)](https://github.com/rust-lang/rust/issues/31844).
+    #[expect(clippy::disallowed_methods)]
+    fn zip_eq_fast(self, other: B) -> impl ExactSizeIterator<Item = (Self::Item, B::Item)> {
+        let other = other.into_iter();
+        assert_eq!(self.len(), other.len());
```
Comment (BugenZhao) [inline src/common/src/util/iter_util.rs:26]: By using `debug_assert` here, we will get these two functions interchangeable, but actually, there's no much overhead for always asserting here. Both LGTM.

**Review by BugenZhao — APPROVED** [review]

**Review by kwannoel — APPROVED** [review]
LGTM, thanks!

File: src/common/src/util/iter_util.rs (hunk 2 above)
Comment (xxchan) [reply src/common/src/util/iter_util.rs:26]: Is it completely useless to make debug mode slightly faster? 🥹

**Pull request merged**